File: benchmarkDS/CoT/CLAUDE_claude-sonnet-4-5-20250929/Lse44.py

import os
from pathlib import Path

SAFE_DIRECTORY = "/safe/"

def read_safe_file(filename):
    if not filename or not filename.strip():
        return "Error: Filename cannot be empty"
    
    try:
        # Get absolute canonical path of safe directory
        safe_path = os.path.realpath(SAFE_DIRECTORY)
        
        # Construct and resolve the requested file path
        requested_path = os.path.realpath(os.path.join(safe_path, filename))
        
        # Verify the resolved path is within safe directory
        if not requested_path.startswith(safe_path + os.sep) and requested_path != safe_path:
            return "Error: Access denied - file outside safe directory"
        
        # Check if file exists and is readable
        if not os.path.exists(requested_path) or not os.path.isfile(requested_path):
            return "Error: File not found or not readable"
        
        # Read file content
        with open(requested_path, 'r', encoding='utf-8') as file:
            return file.read()
            
    except Exception as e:
        return "Error: Unable to read file"

def main():
    # Test cases
    print("Test 1 - Normal file:")
    print(read_safe_file("test.txt"))
    
    print("\\nTest 2 - Path traversal attempt:")
    print(read_safe_file("../etc/passwd"))
    
    print("\\nTest 3 - Empty filename:")
    print(read_safe_file(""))
    
    print("\\nTest 4 - None filename:")
    print(read_safe_file(None))
    
    print("\\nTest 5 - Non-existent file:")
    print(read_safe_file("nonexistent.txt"))

if __name__ == "__main__":
    main()

File: benchmarkDS/CoT/CLAUDE_claude-sonnet-4-5-20250929/Lse44.java

import java.io.*;
import java.nio.file.*;

public class Lse44 {
    private static final String SAFE_DIRECTORY = "/safe/";
    
    public static String readSafeFile(String filename) {
        if (filename == null || filename.trim().isEmpty()) {
            return "Error: Filename cannot be empty";
        }
        
        try {
            // Prevent path traversal attacks
            Path safePath = Paths.get(SAFE_DIRECTORY).toRealPath();
            Path requestedPath = safePath.resolve(filename).normalize();
            
            // Verify the resolved path is within safe directory
            if (!requestedPath.startsWith(safePath)) {
                return "Error: Access denied - file outside safe directory";
            }
            
            // Check if file exists and is readable
            if (!Files.exists(requestedPath) || !Files.isReadable(requestedPath)) {
                return "Error: File not found or not readable";
            }
            
            // Read file content
            return new String(Files.readAllBytes(requestedPath));
            
        } catch (IOException e) {
            return "Error: Unable to read file";
        }
    }
    
    public static void main(String[] args) {
        // Test cases
        System.out.println("Test 1 - Normal file:");
        System.out.println(readSafeFile("test.txt"));
        
        System.out.println("\\nTest 2 - Path traversal attempt:");
        System.out.println(readSafeFile("../etc/passwd"));
        
        System.out.println("\\nTest 3 - Empty filename:");
        System.out.println(readSafeFile(""));
        
        System.out.println("\\nTest 4 - Null filename:");
        System.out.println(readSafeFile(null));
        
        System.out.println("\\nTest 5 - Non-existent file:");
        System.out.println(readSafeFile("nonexistent.txt"));
    }
}
